### PR TITLE
[Merged by Bors] - feat: target of final functor is connected iff source is connected

### PR DESCRIPTION
**Pull request by datokrat**
This PR is part of the effort to prove the Freyd-Mitchell embedding theorem.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by TwoFX** [comment]
Hi, for PRs that build on other PRs, we usually have the later PR (in your case, this one) also have `master` as the target branch, but record the dependency in the PR description. The syntax is explained if you edit the PR description. A bot wil then add the `blocked-by-other-PR` label after a few minutes.

**Comment by alreadydone** [comment]
I think it's okay to make a PR point to a dependent branch in order to reduce diffs, making the PR immediately reviewable without using [another link](https://github.com/leanprover-community/mathlib4/compare/datokrat/FME-100...datokrat/FME-99), but you'd better change the target branch to master before the dependent PR is merged: otherwise GitHub will automatically close this PR once the target branch is merged, and you'll have to restore that branch in order to reopen it and change the target branch.

**Comment by datokrat** [comment]
Thanks for the hints! I'll change the target for now.

File: Mathlib/CategoryTheory/Limits/IsConnected.lean [hunk 1]
```diff
@@ -100,4 +111,15 @@ theorem isConnected_iff_isColimit_pUnitCocone :
   simp only [isConnected_iff_colimit_constPUnitFunctor_iso_pUnit.{w} C]
   exact ⟨colimit.isoColimitCocone colimitCocone⟩
 
+universe v₂ u₂
+variable {C : Type u} {D: Type u₂} [Category.{v} C] [Category.{v₂} D]
+
+/-- The comain of a final functor is connected if and only if its codomain is connected. -/
```
Comment (erdOne) [inline Mathlib/CategoryTheory/Limits/IsConnected.lean:117]:
```suggestion
/-- The domain of a final functor is connected if and only if its codomain is connected. -/
```

**Review by erdOne — APPROVED** [review]
LGTM

File: Mathlib/CategoryTheory/Limits/IsConnected.lean [hunk 2]
```diff
@@ -100,4 +111,15 @@ theorem isConnected_iff_isColimit_pUnitCocone :
...
   exact ⟨colimit.isoColimitCocone colimitCocone⟩
 
+universe v₂ u₂
+variable {C : Type u} {D: Type u₂} [Category.{v} C] [Category.{v₂} D]
+
+/-- The domain of a final functor is connected if and only if its codomain is connected. -/
+theorem isConnected_iff_of_final (F : C ⥤ D) [CategoryTheory.Functor.Final F] :
+    IsConnected C ↔ IsConnected D := by
+  refine Iff.trans (isConnected_iff_colimit_constPUnitFunctor_iso_pUnit.{max v u v₂ u₂} C) ?_
+  refine Iff.trans ?_ (isConnected_iff_colimit_constPUnitFunctor_iso_pUnit.{max v u v₂ u₂} D).symm
```
Comment (joelriou) [inline Mathlib/CategoryTheory/Limits/IsConnected.lean:121]:
```suggestion
  rw [isConnected_iff_colimit_constPUnitFunctor_iso_pUnit.{max v u v₂ u₂} C,
    isConnected_iff_colimit_constPUnitFunctor_iso_pUnit.{max v u v₂ u₂} D]
 ```

File: Mathlib/CategoryTheory/Limits/IsConnected.lean [hunk 3]
```diff
@@ -21,6 +27,11 @@ For connected index categories `C`, `PUnit.{w}` is a colimit of the `constPUnitF
 is arbitrary. See `instHasColimitConstPUnitFunctor`, `isColimitPUnitCocone` and
 `colimitConstPUnitIsoPUnit`.
 
+## Final functors preserve connectedness of categories (in both directions)
+
+`isConnected_iff_of_final` proves that the domain of a final functor is connected if and only if
+  its codomain is connected.
```
Comment (joelriou) [inline Mathlib/CategoryTheory/Limits/IsConnected.lean:33]:
```suggestion
its codomain is connected.
```

**Comment by joelriou** [comment]
Thanks!

bors merge

**Pull request closed without merging**